### ISSUE TITLE
Request // early-exit for already-closed uprobes

### ISSUE THREAD

**Issue by avivzgroundcover** [cilium/ebpf#1013]
Hey all! Back with another strange request for you to consider.

Opening a uprobe (using the link library) registers a finalizer which calls the Close() function. This is understandably needed to make sure no resources are leaked, even if the user did not call the Close() function by themselves.

However - since finalizers are executed sequentelly, when placing a large amount of probes during the program lifetime, process shutdown make take a lot of time. In our usecase, it sometimes takes 30-60 seconds to remove all placed probes (which could be in hundreds of even thousands of probes).

I would like to suggest a simple addition - an early exit condition for the probe's Close function, which checks if closing was already done beforehand.
This will allow the user to manage the shutdown procedure concurrently if speed is of the essence, but still ensure no probes are left open when the process dies.

Thanks!

**Comment by lmb** [comment]
To be clear, Uprobes that you've already `Close()`d have Close invoked again during exit? Or you are after a way to speed up Close during exit?

I'm a bit confused since usually there is no guarantee that finalizers are executed on os.Exit, so it seems like you are doing something custom?

**Comment by ti-mo** [comment]
Hi Aviv,

Have you profiled the shutdown sequence yet? I'm afraid there's not much point discussing this without data.

My wild guess is this should only be noticeable for tracefs events, as closing them requires writing a line to a shared pseudo-file in /proc, and I doubt doing this concurrently is even possible. Double-closing a tracefs event shouldn't be a problem, the second Close() should return quickly with an error.

Have you tried explicitly calling Close() on all probes concurrently?

> an early exit condition for the probe's Close function, which checks if closing was already done beforehand.

Do you mean `perfEvent.Close()`? 

cc @mmat11

**Comment by mmat11** [comment]
> However - since finalizers are executed sequentelly, when placing a large amount of probes during the program lifetime, process shutdown make take a lot of time. In our usecase, it sometimes takes 30-60 seconds to remove all placed probes (which could be in hundreds of even thousands of probes).

does this happen both if you let the finalizer call `link.Close()` and if you `.Close()` concurrently beforehand?

**Comment by avivzgroundcover** [comment]
Sorry for taking long to reply, I appreciate the quick responses!

To clarify - I've noticed a behavior where a program with a large number of uprobes takes a long time to shutdown. I assumed this happens because of the sequential closing of the probes - but I will indeed have to profile the program's shutdown sequence to be more certain. 

About the probe types - this was seen on kernel version 5.10, so these should be PMU probes.

And yes, I have noticed this happening even after closing the probes manually by explicitly calling Close() beforehand. 

I think the best idea is for me to collect further profiling data so it's easier to tell which part takes a long time. 

I will do that and update when I have more news. 

Thanks!
